### DISCUSSION TITLE
How to model all-optical reflectivity changes and trMOKE results starting from udkm1Dsim output

### DISCUSSION THREAD

**Discussion by AvonReppert** [dschick/udkm1Dsim#146] [Ideas]
The udkm1Dsim toolbox is well-suited to model sample responses for a user-defined structure to optical excitation of 1D nanostructures upon fs-laser excitation. It yields spatio-temporal maps of sub-system temperature(s), layer strain, and even sample magnetization based on many common models. In order to compare to experimental observables, it is necessary to also capture the detection process. In that regard, the toolbox offers various possibilities for modeling X-ray diffraction experiments via kinetic and dynamic diffraction theory, even for resonant magnetic scattering conditions.

However, models for the detection process of all-optical methods (i.e., transient reflectivity changes and trMOKE experiments) are not included. All-optical methods are a very common and versatile tool for research in ultrafast dynamics. My question is: What would be a good way forward to model such data based on the output of the udkm1Dsim toolbox (T(z,t), strain(z,t), M(z,t))?

**(1) Case of Reflectivity**
I am aware of this example for modeling all-optical reflectivity changes:
Bojahr, André, et al. "Comparing the oscillation phase in optical pump-probe spectra to ultrafast x-ray diffraction in the metal-dielectric SrRuO3/SrTiO3 superlattice." [Physical Review B 85.22 (2012): 224302.](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.85.224302)

For the case of the reflectivity change, I think of a strain and temperature-dependent refractive index that then goes into the multilayer absorption model already implemented in the toolbox. Then one could calculate the reflectivity for each timestep. This requires knowing the strain and temperature-dependent changes of the complex refractive indices for each material.  This detection process might even be polarization-dependent, and I think only p-polarized light is currently implemented in the formalism. Do you know where I could look for the strain and temperature-dependent changes of the refractive indizes for elemental metals such as (Pt, Ni, Cu, Ta and SiO2?)

**(2) Case of timeresolved magneto-optical Kerr Data**
trMOKE data will be even more complex to model as it involves a polarization dependence. However, the following publication suggests that trMOKE data could also be approximated by a multilayer reflectivity model, but the description is somewhat brief.
Shihab, S., et al. "Counter-rotating standing spin waves: A magneto-optical illusion." [Physical Review B 95.14 (2017): 144411.](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.95.144411)

The referenced publications introduce a sensitivity function for the MOKE detection process that could perhaps be applied to the calculated magnetization map:
Hubert, A., and G. Traeger. "Magneto-optical sensitivity functions of thin-film systems." [Journal of Magnetism and Magnetic Materials 124.1-2 (1993): 185-202.](https://www.sciencedirect.com/science/article/pii/030488539390087I)

Alternatively, what would you suggest as a way forward for modeling the trMOKE response of a laser-excited thin film? I am looking for ideas and related Python packages that could be interfaced with the results of the udkm1Dsim toolbox, which I would use to describe the excitation process. Any ideas?

**Comment by dschick** [comment]
Hi Aleks,

Many thanks for opening this discussion here!
I think it is super stringent to expand the `udkm1Dsim` light scattering simulations also to the visible range.

With exactly this in mind, I have already started to implement the [pyGTM](https://pygtm.readthedocs.io/en/latest/) formalism in PR #97 .
PyGTM should allow to do strain and partly also temperature-dependent simulations even in birefringent materials, which should cover your point **(1)**.  
The current formalism used to calculate the multilayer laser absorption should not differ much.
Temperature-dependent refractive indices should be accessible via DFT while the strain dependence was actually calibrated e.g. in [this paper](https://doi.org10.1103/PhysRevB.86.144306).

Regarding your number **(2)**.
I haven't searched the literature too much for now, so many thanks for the links.
However, we should also discuss if a numerical simulation in a matrix-formalism is really appropriate for most materials.
Also in the optical regime, you will only get notable interference effects at length scales comparable to the light's wavelength.
For most magnetic metals the penetration depth will be much shorter. In YIG this would be much different of course.
Do you have any literature or data on angle-dependent probing of MOKE signals which clearly call for a magneto-optical scattering formalism?

Best

Daniel

**Comment by dschick** [comment]
Hi Aleks,

there might be some progress soon together with some other collaborators.
We found a very nice package named [pyLlama](https://pyllama.readthedocs.io/en/latest/), which we plan to somehow interface with the udkm1Dsim.

Best

Daniel
